pgjdbc/src/main/java/org/postgresql/jdbc/PgPreparedStatement.java

```diff
@@ -682,6 +684,10 @@ public void setObject(@Positive int parameterIndex, @Nullable Object in,
           } else if (in instanceof LocalDateTime) {
             setTimestamp(parameterIndex, (LocalDateTime) in);
             break;
+          } else if (in instanceof Instant) {
+            tmpts = Timestamp.from( (Instant) in );
+          } else if (in instanceof ZonedDateTime) {
+            tmpts = Timestamp.valueOf( ((ZonedDateTime) in).toLocalDateTime());
```

Comment:
Hmm `ZonedDateTime.toLocalDateTime()` would just forget the timezone of the zdt? 
Also, why would it support `ZonedDateTime`, but not `OffsetDateTime`?

I might be wrong with my assumptions here of course.

Reply:
Your assumption is correct, so I changed it to use OffsetDateTime type so that time zone information is not lost but stored into the timestamptz.

Anyway, if you use OffsetDateTime to fetch data back form the db, the time zone info is lost as illustrated by the test case TimestamptzTest.test_ZonedDateTime_StoreAndSelect(). 
How is the correct way to get back the information of the timezone (without using string) ?

NOTE: in our application we always store dates in UTC time zone in the db to avoid any problem. If we need to store information about the time zone to use for some business logic that is stored into a separate fields independent to any stored timestamptz.

Reply:
timestamptz does not store timezone information so there is no way to get it definitely. In text mode it is returned adjusted for the session timezone (most likely client timezone). In binary mode we assume the client timezone.